Mathlib/Data/Greedoid/Predefs.lean

```diff
@@ -0,0 +1,169 @@
+  s₂.card < s₁.card →
+    ∃ x ∈ s₁, ∃ h : x ∉ s₂, cons x s₂ h ∈ Sys
+
+instance {α : Type*} [DecidableEq α] :
+    @DecidablePred (Finset (Finset α)) ExchangeProperty :=
+  fun Sys => by unfold ExchangeProperty; infer_instance
+
+namespace ExchangeProperty
+
+variable {α : Type*} {Sys : Finset (Finset α)} {s₁ s₂ : Finset α}
+
+theorem exists_superset_of_card_le
+  (hSys : ExchangeProperty Sys) (hs₁ : s₁ ∈ Sys) (hs₂ : s₂ ∈ Sys)
+  {n : ℕ} (hn₁ : n ≤ s₁.card) (hn₂ : s₂.card ≤ n) :
+    ∃ s ∈ Sys, s₂ ⊆ s ∧ (∀ ⦃e⦄, e ∈ s → e ∈ s₁ ∨ e ∈ s₂) ∧ s.card = n := by
+  induction n, hn₂ using Nat.le_induction with
+  | base => use s₂; simp [*]; exact fun _ h => .inr h
+  | succ n h ihn =>
+    rcases ihn (by omega) with ⟨t, ht₁, ht₂, ht₃, rfl⟩
+    rcases hSys hs₁ ht₁ hn₁ with ⟨x, _, hx, _⟩
+    use cons x t hx
+    simp [*, subset_trans ht₂ (subset_cons _)]
+    exact fun _ h => ht₃ h
+
+-- TODO: Fix name.
+theorem exists_feasible_superset_add_element_feasible
+  (hSys : ExchangeProperty Sys) (hs₁ : s₁ ∈ Sys) (hs₂ : s₂ ∈ Sys) (hs : s₂ ⊆ s₁)
+  {a : α} (ha₁ : a ∈ s₁) (ha₂ : a ∉ s₂) :
+    ∃ s ∈ Sys, s₂ ⊆ s ∧ s ⊆ s₁ ∧ ∃ h : a ∉ s, cons a s h ∈ Sys := by
+  have h₁ : succ s₂.card ≤ s₁.card := by
+    apply card_lt_card
+    simp only [ssubset_def, hs, true_and]
+    exact fun h => ha₂ (h ha₁)
+  sorry
+
+-- -- TODO: Fix name.
+-- theorem exists_feasible_superset_add_element_feasible
+--     {α : Type _} [DecidableEq α]
+--     {Sys : Finset (Finset α)} (hSys : ExchangeProperty Sys)
+--     {s₁ : Finset α} (hs₁ : s₁ ∈ Sys)
+--     {s₂ : Finset α} (hs₂ : s₂ ∈ Sys)
+--     (hs : s₂ ⊆ s₁)
+--     {a : α} (ha₁ : a ∈ s₁) (ha₂ : a ∉ s₂) :
+--     ∃ s ∈ Sys, s₂ ⊆ s ∧ s ⊆ s₁ ∧ a ∉ s ∧ Insert.insert a s ∈ Sys := by
+--   have h₁ : s₂.card < s₁.card := by
+--     apply card_lt_card
+--     simp only [ssubset_def, hs, true_and]
+--     intro h
+--     exact ha₂ (h ha₁)
+--   by_cases h : Insert.insert a s₂ ∈ Sys
+--   · exists s₂
+--   · let ⟨t, ht₁, ht₂, ht₃, ht₄⟩ := exchangeProperty_exists_superset_of_card_le hSys hs₁ hs₂
+--       (card_le_card hs) h₁ (le_succ _)
+--     have ht₅ : a ∉ t := by
+--       intro h'
+--       apply h; clear h
+--       have h : insert a s₂ = t := by
+--         apply eq_of_subset_of_card_le
+--         · intro _ h
+--           rw [mem_insert] at h
+--           exact h.elim (fun h => h ▸ h') (fun h => ht₂ h)
+--         · rw [ht₄, card_insert_of_not_mem ha₂]
+--       exact h ▸ ht₁
+--     let ⟨s', hs'₁, hs'₂, hs'₃, hs'₄, hs'₅⟩ :=
+--       exchangeProperty_exists_feasible_superset_add_element_feasible hSys hs₁ ht₁
+--         (union_eq_left.mpr hs ▸ ht₃) ha₁ ht₅
+--     exists s'
+--     exact ⟨hs'₁, subset_trans ht₂ hs'₂, hs'₃, hs'₄, hs'₅⟩
+-- termination_by s₁.card - s₂.card
+-- decreasing_by
+--   simp_wf
+--   rw [ht₄]
+--   exact sub_succ_lt_self _ _ h₁
+
+end ExchangeProperty
+
+/-- The accessible property of greedoid -/
+def AccessibleProperty {α : Type*} [DecidableEq α] (Sys : Finset (Finset α)) : Prop :=
```

Comment:
I understand the reason for having `Sys` to be a property on `Finset`s, but what is the reason for having `Sys` itself only apply to finitely many sets, and to itself be a set? It looks to me like `feasible : Finset α → Prop` might be a better type for describing a greedoid. If the feasible sets are all finite, then induction should still work easily. 

Certainly for matroids, it was very convenient to have independence have type `Set α → Prop` rather than `Set (Set α)` or `Set (Finset α)`.

Reply:
I chose to use `Finset` of `Finset`s as used in Mathlib/Combinatorics/SetFamily. I also thought that finiteness of the set family should be explicitly stated. May I ask how `Set α → Prop` has benifit of the implentation while implementing matroids?

Reply:
@apnelson1 I just figured out the file [`Mathlib.Data.Set.Card`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Card.html) which you have submitted last year, and I also think that using finite `Set` with `ncard` should make lives easier! I'll seriously consider porting this project from `Finset` to `Set`, thank you!